Repozo : add an option to verify on recovery

Sebatyne opened the pull request:
This PR aims to reduce the number of IO operations when recovering a ZODB backup with repozo, when user wants to also verify the data. The idea is to add a new option for the -R/--recovery mode which will run the same checks as the -v/--Verify mode, but by reading data in increment files only once.

Also, I did a first commit to prevent the concat() function to close the file descriptor, as currently it is closed twice : once in concat(), and once in the function calling concat(). I think it should be the responsability of the one opening the file to also close it. What surprised me is that there was an assertion in tests checking that concat() was indeed closing the file descriptor. I couldn't find in the git history the explanation why.

mgedmin commented on src/ZODB/scripts/repozo.py:
```diff
@@ -73,6 +73,12 @@
+    --with-verification
+        Verify on the fly the backup files on recovering. This option runs
+        the same checks as when repozo is run in -V/--verify mode, and
+        allows to verify and recover a backup in one single step.
```
I wonder if it's worth mentioning that if verification fails, the output file is left in a partial, inconsistent state.

Sebatyne replied:
That's an interesting issue you are mentioning. From a user experience point of view, do you think that deleting the partially restored zodb would be a good idea ? Otherwise user may use it (ie: after a disaster recovery) without noticinig data is missing. More especially, this partially restored file is valid according to the zodb format.

mgedmin replied:
I was torn about it, and therefore didn't suggest it.  The partial file may be useful for trying to investigate/fix the corruption (although I'm not sure how).  But the risk of the user accidentally using a truncated Data.fs is real.

Perhaps it would be best add a `.tmp` suffix to the output file, and rename it to the final name only if everything's fine.

vpelletier replied:
I like the while-restoring-extension idea. Just to throw in a few more suffix ideas: `.part` ? `.incomplete` ? I am not sure how strong the dot-three extension format rule is these days.

Sebatyne commented:
Hello,

Thank you @icemac and @mgedmin for your good words and your valuable inputs.

I've added an entry in the changelog, and squashed the change into the commit introducing the feature.

As I was at it, I've added a prefix argument to all mkdtemp calls in test_repozo.py, as there weren't numerous.

Regards,

Nicolas

Sebatyne commented on src/ZODB/scripts/repozo.py:
```diff
@@ -730,6 +776,8 @@ def main(argv=None):
         except WouldOverwriteFiles as e:
             sys.exit(str(e))
+        except VerificationFail as e:
+            sys.exit(str(e))
```
Here do you propose to use base class for all 3 repozo custom errors (WouldOverwriteFiles, NoFiles and VerificationFail), or do you propose to merge some of them ? From my understanding, all these exceptions are not raised because of bugs, but just to inform user that repozo didn't finish correctly, most likely because of misuse or protection of users' data.

mgedmin replied:
Yes, a base class for the three custom error classes, with one `except ... as e: sys.exit(str(e))` clause to catch them all in the main function.

I would suggest a name if I could come up with one ;)

Error?  Failure?

icemac commented:
@Sebatyne Thank you for your PR. The Zope Foundation policy requires to sign a Committer agreement before a substantial pull request can be merged to any of its repositories. Please see 
[Zope_Foundation_Committer_Agreement](https://github.com/zopefoundation/developer_docs/raw/master/source/Zope_Foundation_Committer_Agreement-20181008.pdf) for details.

mgedmin commented:
Wow appveyor is deeply unhappy, but the failing test (test_via_monte_carlo) seems to be unrelated to this PR.

If the same test fails on master, I propose we ignore the appveyor failure.

mgedmin commented:
(Rebases/forced pushes make it impossible for me to review just the changes made since I last reviewed the code.  I'm not feeling up to re-reviewing the entire changeset right at the moment.  Please ping me if you think such a review is still needed.)

Sebatyne commented:
Hi,

Sorry for having paused this MR for so long, other things kicked in...

I reviewed one more time the changes today. I've also filled and email the Zope Contributor Agreement.

About Appveyor, I don't have any knowledge about this tool, but it seems a CI suite for windows ? The failing test seems to be caused a file system limitation that I don't have on linux : "[WinError 183] Cannot create a file when that file already exists". As I don't have access to a windows machine, I'll use this PR directly to run my fix attempts on Appveyor.

Regards,

Nicolas

Sebatyne commented:
Hi,

I'll run Appveyor on my fork of ZODB in fact, to avoid spam. I'll contact you again once all tests are OK :)

mgedmin commented on src/ZODB/scripts/repozo.py:
```diff
@@ -649,12 +672,46 @@ def do_recover(options):
+        # old ZODB + full partial file may be superior to free disk space
+        if os.path.exists(options.output):
+          log('Deleting old %s', options.output)
+          os.unlink(options.output)
```
Please use 4 space indents.

Sebatyne commented:
I've squashed the 2 last commits of fixup into the original commit. I'm waiting for tests to run one last time, just in case, then I'll merge.

Thanks to all for your time on this.

icemac commented:
@Sebatyne As you are now a member you should be able to merge this PR yourself.
